File: sdk/typescript/examples/02-caching/sykli.ts
/**
 * Example 02: Input-Based Caching
 *
 * This example demonstrates:
 * - Input patterns for cache keys
 * - Skip execution when inputs unchanged
 * - Multiple input sources
 *
 * Run with: npx ts-node --esm sykli.ts --emit | sykli run -
 */

import { Pipeline, tsInputs } from 'sykli';

const p = new Pipeline();

// Test task with TypeScript inputs
// Only re-runs when .ts/.tsx files or package.json change
p.task('test')
  .run('npm test')
  .inputs(...tsInputs()); // **/*.ts, **/*.tsx, etc.

// Type checking with specific inputs
p.task('typecheck')
  .run('npx tsc --noEmit')
  .inputs('**/*.ts', 'tsconfig.json');

// Lint with custom patterns
p.task('lint')
  .run('npm run lint')
  .inputs('**/*.ts', '**/*.tsx', '.eslintrc*');

// Build depends on all checks passing
p.task('build')
  .run('npm run build')
  .inputs('src/**/*', 'package.json', 'tsconfig.json')
  .after('test', 'typecheck', 'lint');

p.emit();

// With caching, unchanged inputs = instant skip:
//
// $ sykli run
// ✓ test (cached - inputs unchanged)
// ✓ typecheck (cached - inputs unchanged)
// ▶ lint (running...)
// ✓ lint
// ▶ build (running...)
// ✓ build

File: sdk/python/examples/06-matrix/sykli.py
#!/usr/bin/env python3
"""Pipeline with matrix builds across Python versions."""

from sykli import Pipeline

p = Pipeline()

p.task("lint").run("ruff check .")

g = p.matrix(
    "test",
    ["3.11", "3.12", "3.13"],
    lambda v: p.task(f"test-py{v}")
        .run(f"python{v} -m pytest")
        .after("lint"),
)

p.task("build") \
    .run("python -m build") \
    .after_group(g)

p.emit()
